### ISSUE TITLE
Borrow checker thinks variable lives longer than it actually lives

### ISSUE THREAD

**Issue by mosel00** [rust-lang/rust#112087]
The borrow checker thinks that in the following code the variable `v` lives at least as long as `*opt = None;` which is certainly not correct and is also fixed when swapping the two conditions on lines 2 and 3 . To test, [here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&code=fn+f%3C%27a%3E%28opt+%3A+%26%27a+mut+Option%3Ci32%3E%2C+b+%3A+bool%29+-%3E+Result%3C%26%27a+mut+Option%3Ci32%3E%2C+%26%27a+mut+i32%3E+%7B%0A++++if+let+Some%28v%29+%3D+opt+%7B%0A++++++++if+b+%7B%0A++++++++++++return+Err%28v%29%3B%0A++++++++%7D%0A++++%7D%0A%0A++++*opt+%3D+None%3B%0A++++return+Ok%28opt%29%0A%7D%0A%0Afn+main%28%29+%7B%0A++++let+mut+opt+%3D+Some%282%29%3B%0A++++_+%3D+f%28%26mut+opt%2C+false%29%3B%0A%7D)'s a link to the rust playground.

I tried this code:

```rust
fn f<'a>(opt : &'a mut Option<i32>, b : bool) -> Result<&'a mut Option<i32>, &'a mut i32> {
      if let Some(v) = opt {
          if b {
              return Err(v);
          }
      }
    
    *opt = None;
    return Ok(opt)
}
```

I expected to see this happen: no compile error

Instead, this happened: compile error E0506

### Meta
Bug also exists in beta and nightly

`rustc --version --verbose`:
```
binary: rustc
commit-hash: 84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc
commit-date: 2023-04-16
host: x86_64-unknown-linux-gnu
release: 1.69.0
LLVM version: 15.0.7
```

</p>
</details>

**Comment by alexschrod** [comment]
This is a version of #54663

**Comment by fmease** [comment]
Closing as duplicate of #54663.
